src/Workspaces/Remote/ServiceHub/Services/BrokeredServiceBase.cs

```diff
@@ -106,5 +110,39 @@ protected async ValueTask RunServiceAsync(Func<CancellationToken, ValueTask> imp
             cancellationToken = source.Token;
             return source;
         }
+
+        private static bool ReportUnexpectedException(Exception exception, CancellationToken cancellationToken)
+        {
+            if (exception is OperationCanceledException)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return false;
+                }
+
+#if DEBUG
+                // Catch and process an unexpected OCE in the exception handler.
+                // We propagate the OCE in release builds to avoid unnecessary service failures
+                // since most of our code base handles OCE regardless of the token.
+                return FatalError.ReportWithoutCrash(exception);
+#endif
+            }
+
+            return FatalError.ReportWithoutCrashAndPropagate(exception);
+        }
```

Comment:
😕 I don't understand the difference between the debug and release cases. I would expect the caller to just use:

```
when (FatalError.ReportWithoutCrashUnlessCanceledAndPropagate(ex, cancellationToken))
```

Reply:
The debug case catches the exception - the filter returns true and the catch handler wraps the exception in InvalidOperationException. The release just propagates OCE.

Reply:
Can we make debug and release have the same behavior, whichever it is?

Reply:
Yes, we can ignore the OCE in both.